File: PrebidMobile/Rendering/EventHandlers/PrebidMobileMoPubAdapters/Sources/PrebidMoPubBannerAdapter.m
//
//  PBMMoPubBannerAdapter.m
//  OpenXInternalTestApp
//
//  Copyright © 2020 OpenX. All rights reserved.
//
#import <MoPub.h>

#import <PrebidMobileRendering/PBMDisplayView.h>

#import "PrebidMoPubBannerAdapter.h"

/**
 PBM SDK passes to the localExtras two objects: PBMBid, configId
 */

@interface PrebidMoPubBannerAdapter () <PBMDisplayViewLoadingDelegate, PBMDisplayViewInteractionDelegate>

@property (nonatomic, copy) NSString *configId;
@property (strong, nonatomic) PBMDisplayView *displayView;

@end

@implementation PrebidMoPubBannerAdapter

- (void)requestAdWithSize:(CGSize)size adapterInfo:(NSDictionary *)info adMarkup:(NSString *)adMarkup {
    if (self.localExtras.count == 0) {
        NSError *error = [NSError errorWithDomain:PBMErrorDomain
                                             code:PBMErrorCodeGeneral
                                         userInfo:@{NSLocalizedDescriptionKey: NSLocalizedString(@"The local extras is empty", nil)}];
        MPLogAdEvent([MPLogEvent adLoadFailedForAdapter:NSStringFromClass(self.class) error:error], @"");
        [self.delegate inlineAdAdapter:self didFailToLoadAdWithError:error];
        return;
    }
    
    PBMBid *bid = self.localExtras[PBMMoPubAdUnitBidKey];
    if (!bid) {
        NSError *error = [NSError errorWithDomain:PBMErrorDomain
                                             code:PBMErrorCodeGeneral
                                         userInfo:@{NSLocalizedDescriptionKey: NSLocalizedString(@"The Bid object is absent in the extras", nil)}];
        MPLogAdEvent([MPLogEvent adLoadFailedForAdapter:NSStringFromClass(self.class) error:error], [self getAdUnitId]);
        [self.delegate inlineAdAdapter:self didFailToLoadAdWithError:error];
        return;
    }
    
    self.configId = self.localExtras[PBMMoPubConfigIdKey];
    if (!self.configId) {
        NSError *error = [NSError errorWithDomain:PBMErrorDomain
                                             code:PBMErrorCodeGeneral
                                         userInfo:@{NSLocalizedDescriptionKey: NSLocalizedString(@"The Config ID absent in the extras", nil)}];
        MPLogAdEvent([MPLogEvent adLoadFailedForAdapter:NSStringFromClass(self.class) error:error], [self getAdUnitId]);
        [self.delegate inlineAdAdapter:self didFailToLoadAdWithError:error];
        return;
    }
    
    self.displayView = [[PBMDisplayView alloc] initWithFrame:(CGRect){.origin = CGPointZero, .size = bid.size} bid:bid configId:self.configId];
    self.displayView.loadingDelegate = self;
    self.displayView.interactionDelegate = self;
    
    [self.displayView displayAd];
    MPLogAdEvent([MPLogEvent adLoadAttemptForAdapter:NSStringFromClass(self.class) dspCreativeId:nil dspName:nil], [self getAdUnitId]);
}

- (NSString *) getAdUnitId {
    return self.configId ?: @"";
}

#pragma mark - PBMDisplayViewDelegate handlers


- (void)didLeaveAppFromDisplayView:(nonnull PBMDisplayView *)displayView {
    MPLogAdEvent([MPLogEvent adWillLeaveApplicationForAdapter:NSStringFromClass(self.class)], [self getAdUnitId]);
    [self.delegate inlineAdAdapterWillLeaveApplication:self];
}

- (void)displayView:(nonnull PBMDisplayView *)displayView didFailWithError:(nonnull NSError *)error {
    MPLogAdEvent([MPLogEvent adLoadFailedForAdapter:NSStringFromClass(self.class) error:error], [self getAdUnitId]);
    [self.delegate inlineAdAdapter:self didFailToLoadAdWithError:error];
}

- (void)displayViewDidLoadAd:(nonnull PBMDisplayView *)displayView {
    MPLogAdEvent([MPLogEvent adLoadSuccessForAdapter:NSStringFromClass(self.class)], [self getAdUnitId]);
    [self.delegate inlineAdAdapter:self didLoadAdWithAdView:self.displayView];
}

- (void)trackImpressionForDisplayView:(nonnull PBMDisplayView *)displayView {
    //Impressions will be tracked automatically
    //unless enableAutomaticImpressionAndClickTracking = NO
    //In this case you have to override the didDisplayAd method
    //and manually call inlineAdAdapterDidTrackImpression
    //in this method to ensure correct metrics
}

- (nonnull UIViewController *)viewControllerForModalPresentationFrom:(nonnull PBMDisplayView *)displayView {
    return [self.delegate inlineAdAdapterViewControllerForPresentingModalView:self];
}

- (void)displayViewWillPresentModal:(PBMDisplayView *)displayView {
    MPLogAdEvent([MPLogEvent adTappedForAdapter:NSStringFromClass(self.class)], [self getAdUnitId]);
    [self.delegate inlineAdAdapterWillBeginUserAction:self];
}

- (void)displayViewDidDismissModal:(PBMDisplayView *)displayView {
    MPLogInfo(@"Banner's clickthrough did close");
    [self.delegate inlineAdAdapterDidEndUserAction:self];
}

@end

File: PrebidMobile/Rendering/EventHandlers/PrebidMobileMoPubAdapters/Sources/PrebidMoPubBannerAdapter.swift
//
//  PrebidMoPubBannerAdapter.swift
//  PrebidMobileMoPubAdapters
//
//  Copyright © 2021 Prebid. All rights reserved.
//

import Foundation

import MoPubSDK

import PrebidMobileRendering

/**
 Note: Prebid SDK passes to the localExtras two objects: Bid, configId
 */

@objc(PrebidMoPubBannerAdapter)
public class PrebidMoPubBannerAdapter :
    MPInlineAdAdapter,
    PBMDisplayViewLoadingDelegate,
    PBMDisplayViewInteractionDelegate {
   
    // MARK: - Internal Properties
    
    var configID: String?
    var displayView: PBMDisplayView?
    
    // MARK: - MPInlineAdAdapter
    
    public override func requestAd(with size: CGSize,
                                   adapterInfo info: [AnyHashable : Any],
                                   adMarkup: String?) {
        guard !localExtras.isEmpty else {
            let error = MoPubAdaptersError.emptyLocalExtras
            MPLogging.logEvent(MPLogEvent.adLoadFailed(forAdapter: Self.className(), error: error), source: adUnitId, from: nil)
            delegate?.inlineAdAdapter(self, didFailToLoadAdWithError: error)
            return
        }
        
        guard let bid = localExtras[PBMMoPubAdUnitBidKey] as? PBMBid else {
            let error = MoPubAdaptersError.noBidInLocalExtras
            MPLogging.logEvent(MPLogEvent.adLoadFailed(forAdapter: Self.className(), error: error), source: adUnitId, from: nil)
            delegate?.inlineAdAdapter(self, didFailToLoadAdWithError: error)
            return
        }
        
        guard let configID = localExtras[PBMMoPubConfigIdKey] as? String else {
            let error = MoPubAdaptersError.noConfigIDInLocalExtras
            MPLogging.logEvent(MPLogEvent.adLoadFailed(forAdapter: Self.className(), error: error), source: adUnitId, from: nil)
            delegate?.inlineAdAdapter(self, didFailToLoadAdWithError: error)
            return
        }
        
        let frame = CGRect(origin: .zero, size: bid.size)
        displayView = PBMDisplayView(frame: frame,
                                     bid: bid,
                                     configId: configID)
        
        displayView?.loadingDelegate = self
        displayView?.interactionDelegate = self
        
        displayView?.displayAd()
        
        MPLogging.logEvent(MPLogEvent.adLoadAttempt(forAdapter: Self.className(), dspCreativeId: nil, dspName: nil), source: adUnitId, from: nil)
    }
    
    // MARK: - PBMDisplayViewLoadingDelegate
    
    public func displayViewDidLoadAd(_ displayView: PBMDisplayView) {
        MPLogging.logEvent(MPLogEvent.adLoadSuccess(forAdapter: Self.className()), source: adUnitId, from: nil)
        delegate?.inlineAdAdapter(self, didLoadAdWithAdView: self.displayView)
    }
    
    // MARK: - PBMDisplayViewInteractionDelegate
    
    public func displayView(_ displayView: PBMDisplayView, didFailWithError error: Error) {
        MPLogging.logEvent(MPLogEvent.adLoadFailed(forAdapter: Self.className(), error: error), source: adUnitId, from: nil)
        delegate?.inlineAdAdapter(self, didFailToLoadAdWithError: error)
    }
    
    public func trackImpression(for displayView: PBMDisplayView) {
        //Impressions will be tracked automatically
        //unless enableAutomaticImpressionAndClickTracking = NO
        //In this case you have to override the didDisplayAd method
        //and manually call inlineAdAdapterDidTrackImpression
        //in this method to ensure correct metrics
    }
    
    public func viewControllerForModalPresentation(from displayView: PBMDisplayView) -> UIViewController {
        if let controller = delegate?.inlineAdAdapterViewController(forPresentingModalView: self) {
            return controller
        }
        
        // FIXME
        return UIViewController()
    }
    
    public func didLeaveApp(from displayView: PBMDisplayView) {
        // There is no appropriate methods
    }
    
    public func displayViewWillPresentModal(_ displayView: PBMDisplayView) {
        MPLogging.logEvent(MPLogEvent.adTapped(forAdapter: Self.className()), source: adUnitId, from: nil)
        delegate?.inlineAdAdapterWillBeginUserAction(self)
    }
    
    public func displayViewDidDismissModal(_ displayView: PBMDisplayView) {
        MPLogging.logEvent(MPLogEvent.adDidDismissModal(forAdapter: Self.className()), source: adUnitId, from: nil)
        delegate?.inlineAdAdapterDidEndUserAction(self)
    }
}
